File: e2e/tests/discovery.spec.ts
import { test, expect } from '@playwright/test';

test.describe('Discovery', () => {
  test('serves the openid-configuration document', async ({ page }) => {
    await page.goto('/');
    await page.getByRole('button', { name: 'GET openid-configuration' }).click();

    const pre = page.locator('#disc_config');
    await expect(pre).toContainText('"issuer": "dummy"');

    const config = JSON.parse((await pre.textContent())!);
    expect(config.authorization_endpoint).toBe('http://localhost:3000/oauth/authorize');
    expect(config.token_endpoint).toBe('http://localhost:3000/oauth/token');
    expect(config.userinfo_endpoint).toBe('http://localhost:3000/oauth/userinfo');
    expect(config.jwks_uri).toBe('http://localhost:3000/oauth/discovery/keys');
    expect(config.scopes_supported).toContain('openid');
    expect(config.response_types_supported).toEqual(
      expect.arrayContaining(['code', 'id_token', 'id_token token']),
    );
    expect(config.code_challenge_methods_supported).toContain('S256');
    expect(config.id_token_signing_alg_values_supported).toEqual(['RS256']);
    expect(config.subject_types_supported).toEqual(['public']);
  });

  test('serves the JWKS with the RSA public key only', async ({ page }) => {
    await page.goto('/');
    await page.getByRole('button', { name: 'GET jwks (discovery/keys)' }).click();

    const pre = page.locator('#disc_keys');
    await expect(pre).toContainText('"kty": "RSA"');

    const jwks = JSON.parse((await pre.textContent())!);
    expect(jwks.keys).toHaveLength(1);
    const key = jwks.keys[0];
    expect(key.use).toBe('sig');
    expect(key.alg).toBe('RS256');
    expect(key.kid).toBeTruthy();
    expect(key.n).toBeTruthy();
    // Never expose private key material.
    expect(key).not.toHaveProperty('d');
    expect(key).not.toHaveProperty('p');
    expect(key).not.toHaveProperty('q');
  });
});

File: spec/e2e/discovery_spec.rb
# frozen_string_literal: true

require_relative "e2e_helper"

RSpec.describe "Discovery" do
  it "serves the openid-configuration document" do
    visit "/"
    click_button "GET openid-configuration"

    expect(page).to have_css("#disc_config", text: '"issuer": "dummy"')

    config = JSON.parse(find("#disc_config").text)
    expect(config["authorization_endpoint"]).to eq("#{base_url}/oauth/authorize")
    expect(config["token_endpoint"]).to eq("#{base_url}/oauth/token")
    expect(config["userinfo_endpoint"]).to eq("#{base_url}/oauth/userinfo")
    expect(config["jwks_uri"]).to eq("#{base_url}/oauth/discovery/keys")
    expect(config["scopes_supported"]).to include("openid")
    expect(config["response_types_supported"]).to include("code", "id_token", "id_token token")
    expect(config["code_challenge_methods_supported"]).to include("S256")
    expect(config["id_token_signing_alg_values_supported"]).to eq(["RS256"])
    expect(config["subject_types_supported"]).to eq(["public"])
  end

  it "serves the JWKS with the RSA public key only" do
    visit "/"
    click_button "GET jwks (discovery/keys)"

    expect(page).to have_css("#disc_keys", text: '"kty": "RSA"')

    keys = JSON.parse(find("#disc_keys").text)["keys"]
    expect(keys.size).to eq(1)

    key = keys.first
    expect(key["use"]).to eq("sig")
    expect(key["alg"]).to eq("RS256")
    expect(key["kid"]).to be_present
    expect(key["n"]).to be_present
    # Never expose private key material.
    expect(key.keys).not_to include("d", "p", "q")
  end
end
